terraform/README.md

```diff
@@ -0,0 +1,63 @@
+# Terraform module for training-operator
+
+This is a Terraform module facilitating the deployment of the training-operator charm, using the [Terraform juju provider](https://github.com/juju/terraform-provider-juju/). For more information, refer to the provider [documentation](https://registry.terraform.io/providers/juju/juju/latest/docs). 
+
+## Compatibility
+This terraform module is compatible with charms of version >= 1.8 due to changes in the charm's relations.
```

Comment:
Maybe we don't need to provide the whole context. As a reader, if I don't know what relations were in 1.7, I'd probably won't care about this. Perhaps rephrase it to:

```
This terraform module is compatible with the training-operator charm 1.8/stable.
```

I'm hesitant of mentioning any other version as we don't know what comes in the future. It's also going to be an interesting challenge as we generate different versions of a module. Perhaps what we could do is assume that the tf module that lives on the GH branch is compatible only with the charm in the same branch.

Reply:
Keeping the module-charm compatibility in the same branch is an interesting approach. It introduces an initial versioning system too and makes more clear the compatibility pattern. What I 'm worried about though is that:
1. this will introduce the chore of adding modules on every release, since the modules will leave on `track` branches. The scenario of not having new relations is a very likely one for many of our charms and keeping the modules in `main` will mean that we may not have to touch the modules (at least for versioning reasons) for quite a while. 
2. This is a bit scratching the surface of defining the policy of when a module needs to be updated, which we haven't fully defined yet and I 'm afraid that next pulse we may change this if we have a task on this.
3. We have already merged quite a few PRs in `main` :sweat_smile: 

Regarding 1, I think that probably in the future, there will be an automated way of creating those modules since  this looks like something easily automated and maybe maintaining different versions could be done by our CI. But until this is possible, I 'm a bit scared of introducing chores.

Reply:
> The scenario of not having new relations is a very likely one for many of our charms

I think this is really uncertain, it's always a possibility that we need a new relation / deprecate an existing relation. I wouldn't base our versioning system on this assumption.

> We have already merged quite a few PRs in main 😅

If we agree to change this, we can have a small task to update this in the merged PRs.

> there will be an automated way of creating those modules

I fully agree that we can move to automate the versioning chore in the future, but for now seems that it is something we need to maintain.

Another suggestion is to have a less strict wording to explain the actual situation, like
```
This terraform module is compatible with the training-operator charm 1.8 relations. The terraform module can be compatible with >1.8 versions given there are no changes in the relations from 1.8 version.
```
I'm not a big fan of that^ but proposing something that we can build on.

I think mentioning that the relations / the `outputs.tf` file is the one that affects versioning is an important note.

Reply:
> Perhaps what we could do is assume that the tf module that lives on the GH branch is compatible only with the charm in the same branch.

I think this is the most sensible option regarding version support, although it will introduce duplicate code, it follows the same strategy we already use for the actual charms. Hopefully, an automated approach in the future could help speed up the whole process.

> Maybe we don't need to provide the whole context. As a reader, if I don't know what relations were in 1.7, I'd probably won't care about this. 

Regarding this, I think with @DnPlas  in that including potential future versions with `>=` in the compatibility section doesn't make much sense, we should only mention the exact version we have tested on. Including older versions of charms (before 1.8) in the compatibility isn't sensible either, since we don't officially support and never test on them. As a general template, I think this is the most precise:
```
This Terraform module is compatible with the training-operator charm 1.8/stable.
```

As supplemental information, we *could* add a small `Note` below that sentence that is directed at people who know about charms and how they work, mentioning that:
```
Note: The Terraform module may be compatible with other versions of the charm given there are no changes in the relations from version `1.8`.
```

(Also note that the word "Terraform" should always be capitalised per their documentation.

Reply:
As discussed in an online sync (see notes [here](https://docs.google.com/document/d/1VYSCHlfvbGeOKgICOIxZkKTFTdLvll_6BTL8xyGyfGA/edit)), we'll go with:

1. Handling module compatibility in branches, that is, the module in Github branch `track/1.8` of the training operator is compatible with that charm (1.8/stable, 1.8/edge), but the module in branch `main` is compatible with that charm (latest/edge).
2. Removing the note from the README.md as it will be obviated that there is a 1:1 compatibility mapping between the module and the charm in the same Github branch

@orfeas-k, may I ask you to do the appropriate changes?